tsk/fs/encryptionHelper.cpp

```diff
@@ -0,0 +1,181 @@
+* - We found encryption but had an unspecified error in initialization
+* Returns -1 if:
+* - We got far enough to be confident that it's Bitlocker and have a specific error message to get back to the user
+*/
+int handleBitlocker(TSK_FS_INFO* a_fs_info, const char* a_pass) {
+#ifdef HAVE_LIBMBEDTLS
+	BitlockerParser* bitlockerParser = new BitlockerParser();
+	BITLOCKER_STATUS status = bitlockerParser->initialize(a_fs_info->img_info, a_fs_info->offset, a_pass);
+	if (status == BITLOCKER_STATUS::NOT_BITLOCKER) {
+		return 0;
+	}
+
+	if (status != BITLOCKER_STATUS::SUCCESS) {
+
+		// If we have some specific error cases we want to get that information back to the user
+		if (status == BITLOCKER_STATUS::WRONG_PASSWORD) {
+			tsk_error_reset();
+			tsk_error_set_errno(TSK_ERR_FS_BITLOCKER_ERROR);
+			string errStr = "Incorrect password entered " + bitlockerParser->getRecoveryKeyIdStr();
+			tsk_error_set_errstr(errStr.c_str());
+			return -1;
+
+		} else if (status == BITLOCKER_STATUS::NEED_PASSWORD) {
+			tsk_error_reset();
+			tsk_error_set_errno(TSK_ERR_FS_BITLOCKER_ERROR);
+			string errStr = "Password required to decrypt volume " + bitlockerParser->getRecoveryKeyIdStr();
+			tsk_error_set_errstr(errStr.c_str());
+			return -1;
+		}
+		else if (status == BITLOCKER_STATUS::UNSUPPORTED_KEY_PROTECTION_TYPE) {
+			string message = "Unsupported key protection type(s): " + bitlockerParser->getUnsupportedProtectionTypes();
+			tsk_error_reset();
+			tsk_error_set_errno(TSK_ERR_FS_BITLOCKER_ERROR);
+			tsk_error_set_errstr(message.c_str());
+			return -1;
+		}
+		else {
+			// It's unlikely we're going to be able to open the file system (we found at least one BitLocker header) but it's safer to try
+			return 0;
+		}
+
+		delete bitlockerParser;
+		return -1;
+	}
+
+	// Store the BitLocker data to use when reading the volume
+	a_fs_info->encryption_type = TSK_FS_ENCRYPTION_TYPE_BITLOCKER;
+	a_fs_info->encryption_data = (void*)bitlockerParser;
+	a_fs_info->flags |= TSK_FS_INFO_FLAG_ENCRYPTED;
+	a_fs_info->block_size = bitlockerParser->getSectorSize();
+	// We don't set a_fs_info->decrypt_block here because Bitlocker needs to handle both reading in the block
+	// and doing the decryption since some sectors may have been relocated
+#endif
+	return 0;
+}
+
+/**
+* Check if the volume appears to be encrypted and attempt to initialize the encryption object.
+* 
+* @return 0 if:
+* - There was no encryption found
+* - We found encryption and did all the initialization successfully
+* - We found encryption but had an unspecified error in initialization  
+* Returns -1 if:
+* - We found encryption and got far enough that we're confident we should not continue trying to parse the file system and
+*     have potentially useful feedback to give the user (like that the password was incorrect)
+*/
+int handleVolumeEncryption(TSK_FS_INFO* a_fs_info, const char* a_pass) {
+	int ret = 0;
+#ifdef HAVE_LIBMBEDTLS
```

Comment:
Are we ever expecting to not have MBEDTLS? Does it make sense to return 0 if we don't have the library, we can't check so we don't know if there is encryption. Is that the idea that we want to fail safe if we cant check and assume there is no encryption?

Reply:
Resolved I think - mbedtls will not be present in a nolibs build and is not currently included in the linux build (and it would be optional there anyway). And we definitely want to continue trying to open the file system in that case.

Reply:
This can be resolved. This was a lack of understanding on my part.